app/Platform/Controllers/BeatenGamesLeaderboardController.php

```diff
@@ -90,136 +85,213 @@ public function __invoke(Request $request): View
-        }
+            $consoleId = $gameData[$ranking->most_recent_game_id]->ConsoleID ?? null;
+            $ranking->ConsoleName = $consoleId ? $consoleData[$consoleId]->Name ?? null : null;
 
-        if (!$gameKindFilterOptions['prototypes']) {
-            $subquery->where('gd.Title', 'not like', '~Prototype~%');
-        }
+            return $ranking;
+        });
 
-        if (!$gameKindFilterOptions['demos']) {
-            $subquery->where('gd.Title', 'not like', '~Demo~%');
+        // Replace each element in the original collection with its modified version.
+        foreach ($currentPageRows as $key => $modifiedRanking) {
+            $rankingRows[$offset + $key] = $modifiedRanking;
         }
 
-        return $subquery;
+        return $rankingRows;
     }
 
-    private function buildRankingsSubquery(array $gameKindFilterOptions, ?int $targetSystemId = null): mixed
+    private function buildAggregatedLeaderboardQuery(array $gameKindFilterOptions = [], ?int $targetSystemId = null): mixed
     {
-        $subquery = $this->buildLeaderboardBaseSubquery($gameKindFilterOptions, $targetSystemId);
-
-        /** @var string $subqueryTable */
-        $subqueryTable = DB::raw("({$subquery->toSql()}) as s");
-
-        return DB::table($subqueryTable)
-            ->mergeBindings($subquery)
-            ->select(
-                'User',
-                DB::raw('RANK() OVER (ORDER BY COUNT(s.User) DESC) as rank_number'),
-                DB::raw('ROW_NUMBER() OVER (ORDER BY COUNT(s.User) DESC) as leaderboard_row_number'),
-                DB::raw('COUNT(s.User) as total_awards'),
-                'most_recent_game_id',
-                'Title as GameTitle',
-                'ImageIcon as GameIcon',
-                'ConsoleName',
-                DB::raw('MAX(AwardDate) as last_beaten_date'),
-            )
-            ->orderBy('rank_number')
-            ->groupBy('User');
+        $includedTypes = $this->getIncludedTypes($gameKindFilterOptions);
+        $typeBindings = $this->getSubQueryTypeBindings($includedTypes);
+
+        $systemIdCondition = $targetSystemId ? "AND r1.system_id = {$targetSystemId}" : "AND r1.system_id IS NULL";
+        $mostRecentGameIdSubquery = DB::raw("(
+            SELECT r1.game_id
+            FROM rankings AS r1
+            WHERE r1.user_id = rankings.user_id
+                {$systemIdCondition}
+                AND r1.type IN ({$typeBindings})
+            ORDER BY r1.updated_at DESC
+            LIMIT 1
+        ) AS most_recent_game_id");
+
+        $systemIdCondition = $targetSystemId ? "AND r2.system_id = {$targetSystemId}" : "AND r2.system_id IS NULL";
+        $lastBeatenDateSubquery = DB::raw("(
+            SELECT r2.updated_at
+            FROM rankings AS r2
+            WHERE r2.user_id = rankings.user_id
+                {$systemIdCondition}
+                AND r2.type IN ({$typeBindings})
+            ORDER BY r2.updated_at DESC
+            LIMIT 1
+        ) AS last_beaten_date");
```

Comment:
Why does this take two subqueries?
```
> select `user_id`, ( SELECT r1.game_id FROM rankings AS r1 WHERE r1.user_id = rankings.user_id AND r1.system_id IS NULL AND r1.type IN ('games_beaten_hardcore_retail') ORDER BY r1.updated_at DESC LIMIT 1 ) AS most_recent_game_id, ( SELECT r2.updated_at FROM rankings AS r2 WHERE r2.user_id = rankings.user_id AND r2.system_id IS NULL AND r2.type IN ('games_beaten_hardcore_retail') ORDER BY r2.updated_at DESC LIMIT 1 ) AS last_beaten_date, RANK() OVER (ORDER BY SUM(value) DESC) as rank_number, ROW_NUMBER() OVER (ORDER BY SUM(value) DESC) as leaderboard_row_number, SUM(value) as total_awards from `rankings` where `system_id` is null and `type` in ('games_beaten_hardcore_retail') group by `user_id` order by `total_awards` desc, `last_beaten_date` asc
...
39794 rows in set (0.454 sec)
```
vs
```
> SELECT sub.user_id, j.game_id, j.updated_at, sub.total, ROW_NUMBER() OVER(ORDER BY sub.total DESC, sub.updated_at ASC) FROM (SELECT r.user_id, r.system_id, SUM(r.value) AS total, MAX(r.updated_at) AS updated_at FROM rankings r WHERE r.system_id IS NULL AND r.type IN ('games_beaten_hardcore_retail') GROUP BY r.user_id, r.system_id ORDER BY total DESC) AS sub INNER JOIN rankings j ON sub.user_id=j.user_id AND sub.updated_at = j.updated_at AND j.system_id IS NULL ORDER BY sub.total DESC, sub.updated_at ASC
...
39839 rows in set (0.103 sec)
```

Reply:
You're right - using a single subquery in this way is significantly faster. I'm seeing load times go from roughly 500ms to 150ms for the overall leaderboard. The NES leaderboard now also loads in less than 100ms.

It took some tweaking to get the `rank_number` and `leaderboard_row_number` values right, but your example was enough to get me started. Thanks!